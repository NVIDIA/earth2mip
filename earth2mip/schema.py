# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import List, Optional, Mapping, Any
import pydantic
from earth2mip import weather_events
from earth2mip.weather_events import InitialConditionSource, WeatherEvent
from enum import Enum
import datetime
import numpy as np

__all__ = ["InitialConditionSource", "WeatherEvent"]


class Grid(Enum):
    grid_721x1440 = "721x1440"
    grid_720x1440 = "720x1440"
    s2s_challenge = "s2s"

    @property
    def shape(self):
        if self == Grid.grid_721x1440:
            return (721, 1440)
        elif self == Grid.grid_720x1440:
            return (720, 1440)
        elif self == Grid.s2s_challenge:
            return (181, 360)
        else:
            raise ValueError(f"Unknown grid {self}")

    @property
    def _lat_spacing(self):
        if self == Grid.grid_721x1440:
            return 0.25
        elif self == Grid.grid_720x1440:
            return 0.25
        elif self == Grid.s2s_challenge:
            return 1.0
        else:
            raise ValueError(f"Unknown grid {self}")

    @property
    def _lon_spacing(self):
        if self == Grid.grid_721x1440:
            return 0.25
        elif self == Grid.grid_720x1440:
            return 0.25
        elif self == Grid.s2s_challenge:
            return 1.0
        else:
            raise ValueError(f"Unknown grid {self}")

    @property
    def lat(self):
        n = self.shape[0]
        return 90 - np.arange(n) * self._lat_spacing

    @property
    def lon(self):
        n = self.shape[1]
        return np.arange(n) * self._lon_spacing


class InferenceEntrypoint(pydantic.BaseModel):
    """
    Attrs:
        name: an entrypoint string like ``my_package:model_entrypoint``.
            this points to a function ``model_entrypoint(package)`` which returns an
            ``Inference`` object given a package
        kwargs: the arguments to pass to the constructor
    """

    name: str = ""
    kwargs: Mapping[Any, Any] = pydantic.Field(default_factory=dict)


class Model(pydantic.BaseModel):
    """Metadata for using a ERA5 time-stepper model

    Attrs:
        entrypoint: if provided, will be used to load a custom time-loop
            implementation.

    """

    n_history: int = 0
    grid: Grid = Grid.grid_720x1440
    in_channels_names: List[str] = pydantic.Field(default_factory=list)
    out_channels_names: List[str] = pydantic.Field(default_factory=list)
    architecture: str = ""
    architecture_entrypoint: str = ""
    time_step: datetime.timedelta = datetime.timedelta(hours=6)
    entrypoint: Optional[InferenceEntrypoint] = None


class PerturbationStrategy(Enum):
    correlated = "correlated"
    gaussian = "gaussian"
    bred_vector = "bred_vector"
    spherical_grf = "spherical_grf"
    none = "none"


class EnsembleRun(pydantic.BaseModel):
    """A configuration for running an ensemble weather forecast

    Attributes:
        weather_model: The name of the fully convolutional neural network (FCN) model to use for the forecast.
        ensemble_members: The number of ensemble members to use in the forecast.
        noise_amplitude: The amplitude of the Gaussian noise to add to the initial conditions.
        noise_reddening: The noise reddening amplitude, 2.0 was the defualt set by A.G. work.
        simulation_length: The length of the simulation in timesteps.
        output_frequency: The frequency at which to write the output to file, in timesteps.
        use_cuda_graphs: Whether to use CUDA graphs to optimize the computation.
        seed: The random seed for the simulation.
        ensemble_batch_size: The batch size to use for the ensemble.
        autocast_fp16: Whether to use automatic mixed precision (AMP) with FP16 data types.
        perturbation_strategy: The strategy to use for perturbing the initial conditions.
        perturbation_channels: channel(s) perturbed by the initial condition perturbation strategy, None = all channels
        forecast_name (optional): The name of the forecast to use (alternative to `weather_event`).
        weather_event (optional): The weather event to use for the forecast (alternative to `forecast_name`).
        output_dir (optional): The directory to save the output files in (alternative to `output_path`).
        output_path (optional): The path to the output file (alternative to `output_dir`).
        restart_frequency: if provided save at end and at the specified frequency. 0 = only save at end.
        grf_noise_alpha: tuning parameter of the Gaussian random field, see ensemble_utils.generate_noise_grf for details
        grf_noise_sigma: tuning parameter of the Gaussian random field, see ensemble_utils.generate_noise_grf for details
        grf_noise_tau: tuning parameter of the Gaussian random field, see ensemble_utils.generate_noise_grf for details

    """  # noqa

    weather_model: str
    simulation_length: int
    # TODO make perturbation_strategy an Enum (see ChannelSet)
    perturbation_strategy: PerturbationStrategy = PerturbationStrategy.correlated
    perturbation_channels: Optional[List[str]] = None
    noise_reddening: float = 2.0
    noise_amplitude: float = 0.05
    output_frequency: int = 1
    output_grid: Optional[Grid] = None
    ensemble_members: int = 1
    seed: int = 1
    ensemble_batch_size: int = 1
    # alternatives for specifiying forecast
    forecast_name: Optional[str] = None
    weather_event: Optional[weather_events.WeatherEvent] = None
    # alternative for specifying output
    output_dir: Optional[str] = None
    output_path: Optional[str] = None
    restart_frequency: Optional[int] = None
    grf_noise_alpha: float = 2.0
    grf_noise_sigma: float = 5.0
    grf_noise_tau: float = 2.0

    def get_weather_event(self) -> weather_events.WeatherEvent:
        if self.forecast_name:
            return weather_events.read(self.forecast_name)
        else:
            return self.weather_event
