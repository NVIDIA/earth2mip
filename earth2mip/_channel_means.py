# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

channel_means = {
    "u10m": -0.06854962557554245,
    "v10m": 0.18743863701820374,
    "u100m": -0.009113931097090244,
    "v100m": 0.19518691301345825,
    "t2m": 278.7031555175781,
    "sp": 96653.515625,
    "msl": 100957.6796875,
    "tcwv": 18.46604347229004,
    "u1": 10.748488426208496,
    "u2": 9.040291786193848,
    "u3": 7.640656471252441,
    "u5": 6.190047264099121,
    "u7": 5.613990306854248,
    "u10": 4.81683874130249,
    "u20": 3.7416577339172363,
    "u30": 3.9482882022857666,
    "u50": 5.662464618682861,
    "u70": 7.420891761779785,
    "u100": 10.338683128356934,
    "u125": 12.378543853759766,
    "u150": 13.560766220092773,
    "u175": 14.09630012512207,
    "u200": 14.167032241821289,
    "u225": 13.864096641540527,
    "u250": 13.288179397583008,
    "u300": 11.738770484924316,
    "u350": 10.155152320861816,
    "u400": 8.77362060546875,
    "u450": 7.578890800476074,
    "u500": 6.538493633270264,
    "u550": 5.618127822875977,
    "u600": 4.793304920196533,
    "u650": 4.032490253448486,
    "u700": 3.3098866939544678,
    "u750": 2.626377820968628,
    "u775": 2.299865484237671,
    "u800": 1.9830362796783447,
    "u825": 1.6734198331832886,
    "u850": 1.3726757764816284,
    "u875": 1.0908571481704712,
    "u900": 0.8307265639305115,
    "u925": 0.5900797843933105,
    "u950": 0.37076279520988464,
    "u975": 0.17277194559574127,
    "u1000": -0.05516379326581955,
    "v1": -0.081306092441082,
    "v2": -0.021461497992277145,
    "v3": -0.020173462107777596,
    "v5": -0.00731061352416873,
    "v7": -0.004137726966291666,
    "v10": -0.0019085388630628586,
    "v20": -0.00026605380116961896,
    "v30": 0.0010730322683230042,
    "v50": 0.006631731055676937,
    "v70": 0.0045732553116977215,
    "v100": 0.0040822383016347885,
    "v125": -0.02326071448624134,
    "v150": -0.045998428016901016,
    "v175": -0.0552876852452755,
    "v200": -0.03847499191761017,
    "v225": -0.02821849100291729,
    "v250": -0.027320099994540215,
    "v300": -0.02411925420165062,
    "v350": -0.01592441089451313,
    "v400": -0.019759656861424446,
    "v450": -0.026930494233965874,
    "v500": -0.02907356433570385,
    "v550": -0.031823400408029556,
    "v600": -0.02572263590991497,
    "v650": -0.009626002050936222,
    "v700": 0.02697981894016266,
    "v750": 0.06845545023679733,
    "v775": 0.0888051986694336,
    "v800": 0.10753155499696732,
    "v825": 0.12397794425487518,
    "v850": 0.14166374504566193,
    "v875": 0.1662677526473999,
    "v900": 0.19087384641170502,
    "v925": 0.20176121592521667,
    "v950": 0.19572316110134125,
    "v975": 0.18371663987636566,
    "v1000": 0.18469774723052979,
    "w1": -1.1315158872093889e-06,
    "w2": -8.018057087610941e-06,
    "w3": -1.537994648970198e-05,
    "w5": -2.478267197147943e-05,
    "w7": -3.223319072276354e-05,
    "w10": -4.135979543207213e-05,
    "w20": -5.539332050830126e-05,
    "w30": -5.9556386986514553e-05,
    "w50": -5.383044117479585e-05,
    "w70": -4.982736209058203e-05,
    "w100": -2.3971153495949693e-05,
    "w125": 3.87470163332182e-06,
    "w150": 1.66643312695669e-05,
    "w175": 1.5288433132809587e-05,
    "w200": 9.045768820215017e-06,
    "w225": 6.168767413328169e-06,
    "w250": 1.4241597455111332e-05,
    "w300": 9.39343444770202e-05,
    "w350": 0.00019888172391802073,
    "w400": 0.000272307894192636,
    "w450": 0.0002964134037028998,
    "w500": 0.00030192555277608335,
    "w550": 0.0002988489577546716,
    "w600": 0.0003334358334541321,
    "w650": 0.0006918106810189784,
    "w700": 0.002709059976041317,
    "w750": 0.005279306787997484,
    "w775": 0.006850168574601412,
    "w800": 0.008400746621191502,
    "w825": 0.010028586722910404,
    "w850": 0.011793135665357113,
    "w875": 0.01368639525026083,
    "w900": 0.015604125335812569,
    "w925": 0.01744643785059452,
    "w950": 0.019302070140838623,
    "w975": 0.020728887990117073,
    "w1000": 0.020804615691304207,
    "z1": 463206.4375,
    "z2": 411553.125,
    "z3": 382229.46875,
    "z5": 346674.875,
    "z7": 324015.15625,
    "z10": 300542.71875,
    "z20": 256143.234375,
    "z30": 230781.734375,
    "z50": 199379.265625,
    "z70": 179031.015625,
    "z100": 157709.578125,
    "z125": 144308.1875,
    "z150": 133217.546875,
    "z175": 123719.1875,
    "z200": 115399.328125,
    "z225": 107985.0703125,
    "z250": 101282.53125,
    "z300": 89464.4921875,
    "z350": 79181.1875,
    "z400": 70015.5625,
    "z450": 61723.00390625,
    "z500": 54140.7421875,
    "z550": 47151.71484375,
    "z600": 40667.58984375,
    "z650": 34617.66796875,
    "z700": 28943.169921875,
    "z750": 23597.029296875,
    "z775": 21036.0859375,
    "z800": 18544.6875,
    "z825": 16119.4697265625,
    "z850": 13757.439453125,
    "z875": 11455.5458984375,
    "z900": 9210.625,
    "z925": 7019.46533203125,
    "z950": 4878.94189453125,
    "z975": 2786.269775390625,
    "z1000": 738.6001586914062,
    "t1": 261.1324462890625,
    "t2": 255.74183654785156,
    "t3": 247.80210876464844,
    "t5": 237.51869201660156,
    "t7": 231.79122924804688,
    "t10": 226.91737365722656,
    "t20": 219.69107055664062,
    "t30": 216.14366149902344,
    "t50": 212.13992309570312,
    "t70": 209.16539001464844,
    "t100": 208.21510314941406,
    "t125": 210.52633666992188,
    "t150": 213.35560607910156,
    "t175": 215.93296813964844,
    "t200": 218.19093322753906,
    "t225": 220.4385986328125,
    "t250": 222.94039916992188,
    "t300": 229.04507446289062,
    "t350": 235.85696411132812,
    "t400": 242.3114013671875,
    "t450": 248.05335998535156,
    "t500": 253.07688903808594,
    "t550": 257.4590759277344,
    "t600": 261.2414855957031,
    "t650": 264.5358581542969,
    "t700": 267.5036926269531,
    "t750": 270.2417297363281,
    "t775": 271.4933776855469,
    "t800": 272.6558532714844,
    "t825": 273.7077331542969,
    "t850": 274.6662902832031,
    "t875": 275.5856018066406,
    "t900": 276.5234680175781,
    "t925": 277.5076599121094,
    "t950": 278.5635681152344,
    "t975": 279.78955078125,
    "t1000": 281.2162170410156,
    "r1": 0.0012648305855691433,
    "r2": 0.006784914992749691,
    "r3": 0.021754322573542595,
    "r5": 0.10752345621585846,
    "r7": 0.2999992072582245,
    "r10": 0.8303617835044861,
    "r20": 3.4991626739501953,
    "r30": 4.6645050048828125,
    "r50": 6.580825328826904,
    "r70": 13.205595970153809,
    "r100": 26.424592971801758,
    "r125": 26.096942901611328,
    "r150": 26.452449798583984,
    "r175": 29.685396194458008,
    "r200": 35.18281936645508,
    "r225": 41.35318374633789,
    "r250": 46.84218978881836,
    "r300": 53.30179214477539,
    "r350": 53.954071044921875,
    "r400": 52.459781646728516,
    "r450": 51.048095703125,
    "r500": 50.507022857666016,
    "r550": 50.88429260253906,
    "r600": 51.785911560058594,
    "r650": 53.15390396118164,
    "r700": 55.32671356201172,
    "r750": 58.33307647705078,
    "r775": 60.228816986083984,
    "r800": 62.532310485839844,
    "r825": 65.50992584228516,
    "r850": 69.1156234741211,
    "r875": 72.80475616455078,
    "r900": 76.10183715820312,
    "r925": 78.84436798095703,
    "r950": 80.54322052001953,
    "r975": 79.97803497314453,
    "r1000": 77.97394561767578,
    "2d": 274.1190185546875,
    "sst": 189.70262145996094,
    "skt": 279.1072692871094,
    "stl1": 279.7412414550781,
    "swvl1": 0.08473234623670578,
}
