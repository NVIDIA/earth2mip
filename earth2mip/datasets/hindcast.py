# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Routines for opening fcn-mip hindcast outputs
"""
import datetime
import json
import os

import pandas as pd
import xarray
from zarr.storage import FSStore

from earth2mip import filesystem
from earth2mip.datasets.zarr_directory import NestedDirectoryStore


def open_forecast(root, group, chunks=None):
    """Open a fcn-mip forecast as single xarray object

    The directory structure should contain items like this:

        {root}/2018-01-01T00:00:00/mean.zarr/
        {root}/2018-01-02T00:00:00/mean.zarr/

    """
    if isinstance(root, str):
        map_ = FSStore(url=root)
    else:
        map_ = root

    config_path = os.path.join(root, "config.json")
    local_config = filesystem.download_cached(config_path)
    with open(local_config) as f:
        config = json.load(f)
    items = config["protocol"]["times"]

    times = []
    for f in items:
        try:
            datetime.datetime.fromisoformat(f)
        except ValueError:  # noqa
            pass
        else:
            times.append(f)
    times = sorted(times)

    store = NestedDirectoryStore(
        map=map_,
        group=group,
        directories=items,
        concat_dim="initial_time",
        static_coords=("lat", "lon"),
        dim_rename={"time": "lead_time"},
    )

    # TODO this only works locally
    example = xarray.open_zarr(os.path.join(root, f"{items[0]}/{group}"))
    ds = xarray.open_zarr(store, chunks=None).assign_coords(
        {dim: example[dim] for dim in store.static_coords}
    )
    ds["initial_time"] = pd.to_datetime(ds.initial_time)

    if "time" in example.variables:
        ds["lead_time"] = (ds.time - ds.initial_time).isel(initial_time=0)
        ds = ds.rename(time="valid_time", lead_time="time")
    return ds
