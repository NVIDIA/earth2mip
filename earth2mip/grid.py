# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import dataclasses
from typing import List

import numpy as np

from earth2mip import schema


@dataclasses.dataclass(frozen=True)
class LatLonGrid:

    lat: List[float]
    lon: List[float]

    @property
    def shape(self):
        return (len(self.lat), len(self.lon))


def equiangular_lat_lon_grid(
    nlat: int, nlon: int, includes_south_pole: bool = True
) -> LatLonGrid:
    """A regular lat-lon grid

    Lat is ordered from 90 to -90. Includes -90 and only if if includes_south_pole is True.
    Lon is ordered from 0 to 360. includes 0, but not 360.

    """  # noqa
    lat = np.linspace(90, -90, nlat, endpoint=includes_south_pole)
    lon = np.linspace(0, 360, nlon, endpoint=False)
    return LatLonGrid(lat.tolist(), lon.tolist())


def from_enum(grid_enum: schema.Grid) -> LatLonGrid:
    if grid_enum == schema.Grid.grid_720x1440:
        return equiangular_lat_lon_grid(720, 1440, includes_south_pole=False)
    elif grid_enum == schema.Grid.grid_721x1440:
        return equiangular_lat_lon_grid(721, 1440)
    elif grid_enum == schema.Grid.s2s_challenge:
        return equiangular_lat_lon_grid(181, 360)
    else:
        raise ValueError(f"Unknown grid {grid_enum}")
