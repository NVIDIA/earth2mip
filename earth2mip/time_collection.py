# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import datetime
import json
import logging
import shutil
import os

import typer
from modulus.distributed.manager import DistributedManager
import torch.distributed

from distributed import Client

from earth2mip import inference_ensemble, networks, score_ensemble_outputs
from earth2mip.schema import EnsembleRun

logging.basicConfig(
    format="%(asctime)s:%(levelname)-s:%(name)s:%(message)s",
    level=logging.INFO,
    datefmt="%Y-%m-%d %H:%M:%S",
)

logger = logging.getLogger(__file__)
WD = os.getcwd()


def get_distributed_client(rank):
    scheduler_file = "scheduler.json"
    if rank == 0:
        client = Client(n_workers=32, threads_per_worker=1)
        client.write_scheduler_file(scheduler_file)

    if torch.distributed.is_initialized():
        torch.distributed.barrier()

    if rank != 0:
        client = Client(scheduler_file=scheduler_file)

    return client


def main(
    root: str,
    shard: int = 0,
    n_shards: int = 1,
):
    """
    Args:
        root: the root directory of the output
        shard: index of the shard
        n_shards: split the input times into this many shards
    """
    assert shard < n_shards

    time = datetime.datetime(1, 1, 1)

    config_path = os.path.join(root, "config.json")
    with open(config_path, "r") as f:
        config = json.load(f)

    DistributedManager.initialize()
    model = networks.get_model(config["model"])
    dist = DistributedManager()

    protocol = config["protocol"]
    lines = protocol["times"][shard::n_shards]
    logger.info(
        f"Working on shard {shard+1}/{n_shards}. {len(lines)} initial times to run."
    )

    run = EnsembleRun.parse_obj(protocol["inference_template"])
    n_ensemble_batches = run.ensemble_members // run.ensemble_batch_size
    ranks_per_time = min(n_ensemble_batches, dist.world_size)
    ranks_per_time = ranks_per_time - dist.world_size % ranks_per_time

    time_rank = int(dist.rank // ranks_per_time)
    n_time_groups = int(dist.world_size // ranks_per_time)

    group_ranks = list(
        range(time_rank * ranks_per_time, (time_rank + 1) * ranks_per_time)
    )
    logger.info(
        "distributed info: " + str((dist.rank, time_rank, group_ranks, ranks_per_time))
    )

    if torch.distributed.is_initialized():
        group = torch.distributed.new_group(group_ranks)
        group_rank = torch.distributed.get_group_rank(group, dist.rank)
        lines = lines[time_rank::n_time_groups]

    # setup dask client for post processing
    client = get_distributed_client(dist.rank)
    post_process_task = None

    count = 0
    for line in lines:
        count += 1
        initial_time = datetime.datetime.fromisoformat(line.strip())
        start = time.now()
        output = f"{root}/{initial_time.isoformat()}"
        if os.path.exists(output):
            continue

        d = output + ".tmp"

        if torch.distributed.is_initialized():
            torch.distributed.barrier(group)

        perturb = inference_ensemble.get_initializer(
            model,
            run,
        )
        run.weather_event.properties.start_time = initial_time
        run.output_path = d
        inference_ensemble.run_inference(
            model, run, group=group, progress=False, perturb=perturb
        )

        if group_rank == 0:

            def post_process(d):
                output_path = f"{d}/output/"
                score_ensemble_outputs.main(
                    input_path=d,
                    output_path=output_path,
                    time_averaging_window=protocol.get("time_averaging_window", ""),
                    score=protocol.get("score", False),
                    save_ensemble=protocol.get("save_ensemble", False),
                )
                shutil.move(output_path, output)
                shutil.rmtree(d, ignore_errors=True)

            # do not work on more than one post processing job at once
            if post_process_task is not None:
                post_process_task.result()

            post_process_task = client.submit(post_process, d)

            stop = time.now()
            elapsed = stop - start
            remaining = elapsed * (len(lines) - count)
            logger.info(
                f"{count}/{len(lines)}: {initial_time} done. Elapsed: {elapsed.total_seconds()}s. Remaining: {remaining}s"  # noqa
            )

    # finish up final task
    if group_rank == 0 and post_process_task is not None:
        post_process_task.result()

    # keep barrier at end so
    # dask distributed client is not cleaned up
    if torch.distributed.is_initialized():
        torch.distributed.barrier()

    logging.info(f"rank {dist.rank} Finished.")


if __name__ == "__main__":
    typer.run(main)
