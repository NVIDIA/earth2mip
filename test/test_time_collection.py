# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import datetime
import json
import os
import subprocess

import numpy as np
import pytest
import torch
import xarray

import earth2mip.grid
import earth2mip.schema
from earth2mip.datasets.hindcast import open_forecast
from earth2mip.time_collection import run_over_initial_times

DIR = os.path.dirname(__file__)
os.environ["PYTHONPATH"] = DIR + ":" + os.getenv("PYTHONPATH", ":")


def create_model_package(tmp_path):
    package_dir = tmp_path / "mock_package"
    package_dir.mkdir()

    # Create metadata.json
    metadata = {
        "architecture_entrypoint": "mock_plugin:load",
        "n_history": 0,
        "grid": "721x1440",
        "in_channels": list(range(73)),
        "out_channels": list(range(73)),
    }
    with open(package_dir.joinpath("metadata.json"), "w") as f:
        json.dump(metadata, f)

    # Create numpy arrays
    global_means = np.zeros((1, 73, 1, 1))
    global_stds = np.ones((1, 73, 1, 1))
    np.save(package_dir.joinpath("global_means.npy"), global_means)
    np.save(package_dir.joinpath("global_stds.npy"), global_stds)

    return package_dir


@pytest.mark.slow
@pytest.mark.xfail
def test_time_collection(tmp_path):
    if not torch.cuda.is_available():
        pytest.skip("needs gpu and data")

    torch.init_process_group()
    model_package = create_model_package(tmp_path)

    config = os.path.join(DIR, "configs/medium-test.json")
    model = f"file://{model_package.as_posix()}"

    root = str(tmp_path / "test")
    subprocess.check_call(
        ["python3", "-m", "earth2mip.make_job", model, config, root]  # noqa: S603 S607
    )  # noqa: S603 S607
    subprocess.check_call(  # noqa: S603 S607
        [  # noqa: S603 S607
            "torchrun",
            "--nnodes",
            "1",
            "--nproc_per_node",
            str(min([torch.cuda.device_count(), 2])),
            "-m",
            "earth2mip.time_collection",
            root,
        ]
    )  # noqa: S603 S607

    ds = open_forecast(root, group="mean.zarr")
    assert isinstance(ds, xarray.Dataset)


@pytest.mark.slow
def test_run_over_initial_times(tmp_path, regtest):
    class MockTimeLoop:
        in_channel_names = ["b", "a"]
        out_channel_names = ["b", "a"]
        time_step = datetime.timedelta(hours=6)
        history_time_step = datetime.timedelta(hours=6)
        n_history_levels = 1
        grid = earth2mip.grid.equiangular_lat_lon_grid(2, 2)
        device = "cpu"
        dtype = torch.float

        def __call__(self, time, x):
            while True:
                yield time, x[:, 0], None
                time += self.time_step

    class MockData:
        grid = MockTimeLoop.grid
        channel_names = MockTimeLoop.in_channel_names

        def __getitem__(self, time):
            return np.zeros([len(self.channel_names), *self.grid.shape])

    # an template configuration for a short ensemble run
    config_obj = {
        "ensemble_members": 4,
        "simulation_length": 10,
        "weather_event": {
            "properties": {"name": "global", "start_time": "2018-01-01T00:00:00"},
            "domains": [
                {
                    "name": "global",
                    "type": "Window",
                    "diagnostics": [
                        {
                            "type": "raw",
                            "channels": MockTimeLoop.out_channel_names,
                        }
                    ],
                }
            ],
        },
        "output_path": "dummy/output",
        "output_frequency": 1,
        "weather_model": "unused",
        "seed": 12345,
        "use_cuda_graphs": False,
        "ensemble_batch_size": 1,
        "autocast_fp16": False,
        "perturbation_strategy": "correlated",
        "noise_amplitude": 0.05,
        "noise_reddening": 2,
    }

    time_loop = MockTimeLoop()
    data_source = MockData()
    initial_times = [datetime.datetime(2018, 1, 1), datetime.datetime(2018, 1, 2)]
    config = earth2mip.schema.EnsembleRun.parse_obj(config_obj)
    output_path = tmp_path.as_posix()

    run_over_initial_times(
        time_loop=time_loop,
        data_source=data_source,
        initial_times=initial_times,
        config=config,
        output_path=output_path,
        n_post_processing_workers=1,
    )

    # open the data and ensure its format is the same as the regression data
    ds = open_forecast(output_path, "mean.zarr")
    # print the CDL-like schema of the dataset to the regression fixture
    # ``regtest``
    # delete some non-reproducible attrs
    del ds.attrs["date_created"]
    del ds.attrs["history"]
    ds.info(regtest)
