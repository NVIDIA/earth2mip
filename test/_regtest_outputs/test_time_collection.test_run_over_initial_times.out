xarray.Dataset {
dimensions:
	initial_time = 2 ;
	time = 11 ;
	lat = 2 ;
	lon = 2 ;

variables:
	float64 a(initial_time, time, lat, lon) ;
	float64 b(initial_time, time, lat, lon) ;
	float64 imax(initial_time) ;
	float64 imin(initial_time) ;
	datetime64[ns] initial_time(initial_time) ;
	float64 jmax(initial_time) ;
	float64 jmin(initial_time) ;
	datetime64[ns] valid_time(initial_time, time) ;
	float32 lat(lat) ;
		lat:long_name = latitude ;
		lat:standard_name = latitude ;
		lat:units = degrees_north ;
	float32 lon(lon) ;
		lon:long_name = longitude ;
		lon:standard_name = longitude ;
		lon:units = degrees_east ;
	timedelta64[ns] time(time) ;

// global attributes:
	:Conventions = CF-1.10 ;
	:config = {"weather_model": "unused", "simulation_length": 10, "perturbation_strategy": "correlated", "perturbation_channels": null, "noise_reddening": 2.0, "noise_amplitude": 0.05, "output_frequency": 1, "output_grid": null, "ensemble_members": 4, "seed": 12345, "ensemble_batch_size": 1, "forecast_name": null, "weather_event": {"properties": {"name": "global", "start_time": "2018-01-01T00:00:00", "initial_condition_source": "era5", "netcdf": "", "restart": ""}, "domains": [{"type": "Window", "name": "global", "lat_min": -90, "lat_max": 90, "lon_min": 0, "lon_max": 360, "diagnostics": [{"type": "raw", "function": "", "channels": ["b", "a"], "nbins": 10}]}]}, "output_dir": null, "output_path": "<pytest_tempdir>/test_run_over_initial_times0/2018-01-01T00:00:00.tmp", "restart_frequency": null, "grf_noise_alpha": 2.0, "grf_noise_sigma": 5.0, "grf_noise_tau": 2.0} ;
	:institution = NVIDIA ;
	:model = unused ;
	:time_averaging_window =  ;
	:weather_event = {"properties": {"name": "global", "start_time": "2018-01-01T00:00:00", "initial_condition_source": "era5", "netcdf": "", "restart": ""}, "domains": [{"type": "Window", "name": "global", "lat_min": -90, "lat_max": 90, "lon_min": 0, "lon_max": 360, "diagnostics": [{"type": "raw", "function": "", "channels": ["b", "a"], "nbins": 10}]}]} ;
	:_ARRAY_DIMENSIONS = ['initial_time'] ;
}